File: DataStructsAndAlgos/implementations/python/src/ds/Trees/bst.py
from .. Stack.stack import Stack
from .. Queue.queue import Queue
from .. util.trees import TreeNode


class BST:
    def __init__(self, elem):
        self.root = TreeNode(elem)

    def init(self, newElem):
        if self.root == None:
            self.root = TreeNode(newElem)
            return True
        elif self.root.left == None and newElem < self.root.elem:
            self.root.left = TreeNode(newElem)
            return True
        elif self.root.right == None and newElem >= self.root.elem:
            self.root.right = TreeNode(newElem)
            return True
        return False

    # traverse returns a tuple
    # left element represents previous pointer to the one you want to find
    # right element represents pointer to the element you are looking for

    def traverse(self, func, elemToFind):
        currentPtr = self.root
        prevPtr = currentPtr
        while currentPtr != None:
            if func.upper() == 'SEARCH' and currentPtr.elem == elemToFind:
                break
            prevPtr = currentPtr
            if elemToFind >= currentPtr.elem:
                currentPtr = currentPtr.right
            else:
                currentPtr = currentPtr.left
        return (prevPtr, currentPtr)

    def destroy(self):
        if self.root.left == None and self.root.right == None:
            ptr = self.root
            self.root = None
            del ptr
            return True
        return False

    def insert(self, newElem):
        if not self.init(newElem):
            ptrs = self.traverse('', newElem)
            if newElem >= ptrs[0].elem:
                ptrs[0].right = TreeNode(newElem)
            else:
                ptrs[0].left = TreeNode(newElem)

    def removeElem(self, newElem):
        ptrs = self.traverse('SEARCH', newElem)
        if ptrs[1] == None:
            return None
        elif ptrs[1].left == None and ptrs[1].right == None:
            if ptrs[0].left == ptrs[1]:
                ptrs[0].left = None
            else:
                ptrs[0].right = None
        elif ptrs[1].left == None or ptrs[1].right == None:
            nextPtr = ptrs[1]
            if ptrs[1].right != None:
                nextPtr = nextPtr.right
                ptrs[1].right = None
            else:
                nextPtr = nextPtr.left
                ptrs[1].left = None
            ptrs[1].elem = nextPtr.elem
        else:
            ptrToReplace = ptrs[1]
            ptrToPrecedeDelete = ptrToReplace
            ptrToDelete = ptrToPrecedeDelete.left
            while ptrToDelete.right != None:
                ptrToPrecedeDelete = ptrToDelete
                ptrToDelete = ptrToDelete.right
            ptrToReplace.elem = ptrToDelete.elem
            if ptrToDelete.left != None:
                ptrToPrecedeDelete.left = ptrToDelete.left
                ptrToDelete.left = None
            elif ptrToPrecedeDelete.left == ptrToDelete:
                ptrToPrecedeDelete.left = None
            else:
                ptrToPrecedeDelete.right = None

    def binarySearch(self, elemToFind):
        ptrs = self.traverse(self, elemToFind)
        if ptrs[1] != None and ptrs[1].elem == elemToFind:
            return True
        return False

    def PreOrderRecursive(self, root=None, preOrder=[]):
        newRoot = root
        if root == None:
            newRoot = self.root
        preOrder.append(newRoot.elem)
        self.PreOrderRecursive(newRoot.left, preOrder)
        self.PreOrderRecursive(newRoot.right, preOrder)

    def InOrderRecursive(self, elem):
        pass

    def PostOrderRecursive(self, elem):
        pass

    def PreOrderIterative(self, elem):
        pass

    def InOrderIterative(self, elem):
        pass

    def DFS(self, elemToFind):
        pass

    def BFS(self, elemToFind):
        pass

    def BFSTraversal(self):
        BFSTraversal = []
        BFSQueue = Queue(self.root)
        while(BFSQueue.getLength() > 0):
            newRoot = BFSQueue.pop()
            BFSTraversal.append(newRoot.elem)
            if newRoot.left != None:
                BFSQueue.push(newRoot.left)
            if newRoot.right != None:
                BFSQueue.push(newRoot.right)
        return BFSTraversal

File: DataStructsAndAlgos/implementations/javascript/src/DataStructures/Trees/BST/BST.js
import { Stack } from '../../Stack/Stack';
import { Queue } from '../../Queue/Queue';
import { TreeNode } from '../TreeNode';

export class BST {
  constructor(elem) {
    this.root = null;
    if (!!elem) this.init(new TreeNode(elem));
  }

  init(newNode) {
    this.root = newNode;
  }
  cleanup() {
    this.root = null;
  }
  traverseToBottom(newElem, current, next) {
    while (!!next) {
      current = next;
      if (newElem >= next.elem) next = next.right;
      else next = next.left;
    }
  }
  traverseToFind(elemToFind, current, prev) {
    while (!!current && current.elem !== elemToFind) {
      prev = current;
      if (elemToFind >= current.elem) current = elemToRemove.right;
      else current = current.left;
    }
  }
  traverseToLargestSubTree(ptr, prevPtr) {
    while (!!ptr.right && !!ptr.right.right) {
      prevPtr = ptr;
      ptr = ptr.right;
    }
  }
  insert(newElem) {
    let newNode = new TreeNode(newElem);
    if (this.root === null) this.init(newNode);
    else if (!this.root.left && newElem < this.root.elem)
      this.root.left = newNode;
    else if (!this.root.right && newElem >= this.root.elem)
      this.root.right = newNode;
    else {
      let ptrs = { current: this.root, next: this.root };
      let current = this.root;
      let next = this.root;

      //TODO: Refactor into traverse method.
      while (!!next) {
        current = next;
        if (newElem >= next.elem) next = next.right;
        else next = next.left;
      }

      if (newElem >= current.elem && !current.right) current.right = newNode;
      else current.left = newNode;
    }
  }
  removeByElem(elem) {
    if (this.root.left === null && this.root.right === null) this.cleanup();
    else {
      let prev = this.root;
      let elemToRemove = this.root;
      //TODO: Refactor into traverse method
      while (!!elemToRemove && elemToRemove.elem !== elem) {
        prev = elemToRemove;
        if (elem >= elemToRemove.elem) elemToRemove = elemToRemove.right;
        else elemToRemove = elemToRemove.left;
      }
      if (!elemToRemove.left && !elemToRemove.right) {
        //case 0: elementToRemove has no children
        //figure out which side elemToRemove is from prev
        if (!!prev.right && prev.right.elem === elemToRemove.elem)
          prev.right = null;
        else prev.left = null;
      } else if (!elemToRemove.left || !elemToRemove.right) {
        prev.elem = elemToRemove.elem;
        //case 1: elemToRemove has ONLY a right child
        if (!elemToRemove.left && !!elemToRemove.right) prev.right = null;
        //case 2: elemToRemove has ONLY a left child
        else prev.left = null;
      } else {
        //case 3: elemToRemove has both a left child and a right child
        let newPtr = elemToRemove;
        elemToRemove = elemToRemove.left;
        prev = newPtr;
        //TODO: Refactor into traverse method
        while (!!elemToRemove.right) {
          prev = elemToRemove;
          elemToRemove = elemToRemove.right;
        }

        newPtr.elem = elemToRemove.elem;
        if (!!elemToRemove.left) {
          if (prev.left === elemToRemove) prev.left = elemToRemove.left;
          else prev.right = elemToRemove.left;
          elemToRemove.left = null;
        } else {
          if (prev.left === elemToRemove) prev.left = null;
          else prev.right = null;
        }
      }
    }
  }
  //Traversals
  PreOrderRecursive(returnArray, root = this.root) {
    if (!root) return;
    returnArray.push(root.elem);
    this.PreOrderRecursive(returnArray, root.left);
    this.PreOrderRecursive(returnArray, root.right);
  }
  InOrderRecursive(returnArray, root = this.root) {
    if (!root) return;
    this.InOrderRecursive(returnArray, root.left);
    returnArray.push(root.elem);
    this.InOrderRecursive(returnArray, root.right);
  }
  PostOrderRecursive(returnArray, root = this.root) {
    if (!root) return;
    this.PostOrderRecursive(returnArray, root.left);
    this.PostOrderRecursive(returnArray, root.right);
    returnArray.push(root.elem);
  }
  PreOrderIterative() {}
  InOrderIterative() {}
  //Searches
  BinarySearchIterative(elem) {
    let root = this.root;
    while (root !== null) {
      if (root.elem === elem) return true;
      else if (elem >= root.elem) root = root.right;
      else root = root.left;
    }
    return false;
  }
  BinarySearchRecursive() {}
  SearchIterative(search) {
    let output = [];
    let SearchStructure;
    switch (search.toUpperCase()) {
      case 'BFS':
        SearchStructure = new Queue();
        break;
      case 'DFS':
        SearchStructure = new Stack();
        break;
      default:
        return [];
    }
    SearchStructure.push(this.root);
    while (SearchStructure.length() > 0) {
      let root = SearchStructure.pop();
      output.push(root.elem);
      if (search.toUpperCase() === 'BFS') {
        if (!!root.left) SearchStructure.push(root.left);
        if (!!root.right) SearchStructure.push(root.right);
      } else if (search.toUpperCase() === 'DFS') {
        if (!!root.right) SearchStructure.push(root.right);
        if (!!root.left) SearchStructure.push(root.left);
      }
    }
    return output;
  }

  BFSIterative() {}
  BFSRecursive() {}

  DFSRecursive() {}
}
